lib/build-metadata.ts

```diff
@@ -0,0 +1,129 @@
+import { Either, isLeft } from 'fp-ts/lib/Either';
+import * as t from 'io-ts';
+import { PathReporter } from 'io-ts/lib/PathReporter';
+import * as jsYaml from 'js-yaml';
+import * as _ from 'lodash';
+import * as Stream from 'stream';
+import * as tar from 'tar-stream';
+import * as TarUtils from 'tar-utils';
+
+import { BalenaYml, parsedBalenaYml, ParsedBalenaYml } from './build-secrets';
+import { BalenaYMLValidationError } from './errors';
+import * as PathUtils from './path-utils';
+
+enum MetadataFileType {
+	Json,
+	Yaml,
+}
+
+export class BuildMetadata {
+	private metadataFiles: Dictionary<Buffer> = {};
+	private balenaYml: BalenaYml;
+
+	public constructor(private metadataDirectory: string) {}
+
+	public async extractMetadata(
+		tarStream: Stream.Readable,
+	): Promise<Stream.Readable> {
+		// Run the tar file through the extraction stream, removing
+		// anything that is a child of the metadata directory
+		// and storing it, otherwise forwarding the other files to
+		// a tar insertion stream, which is then returned.
+		const onEntry = async (
+			pack: tar.Pack,
+			header: tar.Headers,
+			stream: Stream.Readable,
+		) => {
+			const buffer = await TarUtils.streamToBuffer(stream);
+			const relative = this.getMetadataRelativePath(header.name);
+
+			if (relative != null) {
+				this.addMetadataFile(relative, buffer);
+			} else {
+				pack.entry(header, buffer);
+			}
+		};
+		return (await TarUtils.cloneTarStream(tarStream, {
+			onEntry,
+		})) as Stream.Readable;
+	}
+
+	public getBalenaYml() {
+		return this.balenaYml;
```

Comment:
This allows modifying `this.balenaYml`. Maybe it should return a copy?

Reply:
Probably worth doing, as it is though, it only gets used internally.